### PR TITLE
Fixes recycler factory provider initialization

### PR DESCRIPTION
**Pull request by vy**
Discovered while porting #2533 to `main`.

### CODE REVIEW

**Comment by ppkarwasz** [comment]
I think that the `RecyclerFactoryProvider.createForEnvironment` method should never return `null`.

If we want to deal with the specific case of `ThreadLocalRecyclerFactoryProvider` that should be disabled by **default** if `log4j2.enableThreadlocals` is set to `false`, we can play with its order:

https://github.com/apache/logging-log4j2/blob/8d19bc5e84ca74d01bc5e1ad7731d1b0063419f2/log4j-kit/src/main/java/org/apache/logging/log4j/kit/recycler/internal/ThreadLocalRecyclerFactoryProvider.java#L48-L51

If the order is high enough, the dummy provider will prevail unless the user **explicitly** requested `threadLocal` to be used.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/impl/CoreDefaultBundle.java [hunk 1]
```diff
@@ -103,24 +105,58 @@ public FlowMessageFactory defaultFlowMessageFactory() {
...
-        return provider.createForEnvironment(environment);
+        // Collect providers
+        @Nullable
+        final String providerName =
+                environment.getProperty(RecyclerProperties.class).factory();
+        final List<RecyclerFactoryProvider> providers = ServiceLoaderUtil.safeStream(
+                        RecyclerFactoryProvider.class,
+                        ServiceLoader.load(RecyclerFactoryProvider.class, loader),
+                        statusLogger)
+                .sorted(Comparator.comparing(RecyclerFactoryProvider::getOrder))
```
Comment (ppkarwasz) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/impl/CoreDefaultBundle.java:121]: Can we use a composite comparator such that:

1. The provider supplied using `log4j.recycler.factory` comes first,
2. The remaining providers are ordered by `Order`.

```java
.sorted(Comparator.comparing(p -> p.getName().equals(providerName) ? Integer.MIN_VALUE : p.getOrder()
```

should work

**Comment by vy** [comment]
> I think that the `RecyclerFactoryProvider.createForEnvironment` method should never return `null`.

Right. Fixed in b741632db09ad8cecb59bdf78a46eebfb0ac06c9.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/impl/CoreDefaultBundle.java (hunk 1 above)
Comment (vy) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/impl/CoreDefaultBundle.java:121]: I prefer a more dumb-proof style: Have an explicit block for `if (providerName != null)`.

**Pull request merged**